components/tidb_query/src/rpn_expr/impl_math.rs

```diff
@@ -147,42 +214,118 @@ impl Floor for FloorIntToInt {
     }
 }
 
-#[rpn_fn]
 #[inline]
-fn abs_int(arg: &Option<Int>) -> Result<Option<Int>> {
-    match arg {
-        None => Ok(None),
-        Some(arg) => match (*arg).checked_abs() {
-            None => Err(Error::overflow("BIGINT", &format!("abs({})", *arg)).into()),
-            Some(arg_abs) => Ok(Some(arg_abs)),
-        },
+#[rpn_fn]
+pub fn round<R: Round>(arg: &Option<R::Type>) -> Result<Option<R::Type>> {
+    if let Some(arg) = arg {
+        R::round(arg)
+    } else {
+        Ok(None)
     }
 }
 
-#[rpn_fn]
 #[inline]
-fn abs_uint(arg: &Option<Int>) -> Result<Option<Int>> {
-    Ok(*arg)
+#[rpn_fn]
+pub fn round_with_frac<R: Round>(
+    arg0: &Option<R::Type>,
+    arg1: &Option<Int>,
+) -> Result<Option<R::Type>> {
+    match (arg0, arg1) {
+        (Some(number), Some(digits)) => R::round_with_frac(number, *digits),
+        _ => Ok(None),
+    }
 }
 
-#[rpn_fn]
-#[inline]
-fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
-    match arg {
-        Some(arg) => Ok(Some(num_traits::Signed::abs(arg))),
-        None => Ok(None),
+pub trait Round {
+    type Type: Evaluable;
+
+    fn round(_arg: &Self::Type) -> Result<Option<Self::Type>> {
+        Ok(None)
+    }
+    fn round_with_frac(_number: &Self::Type, _digits: Int) -> Result<Option<Self::Type>> {
+        Ok(None)
     }
 }
 
-#[rpn_fn]
-#[inline]
-fn abs_decimal(arg: &Option<Decimal>) -> Result<Option<Decimal>> {
-    match arg {
-        Some(arg) => {
-            let res: codec::Result<Decimal> = arg.to_owned().abs().into();
-            Ok(Some(res?))
+pub struct RoundInt;
+
+impl Round for RoundInt {
+    type Type = Int;
+
+    #[inline]
+    fn round(arg: &Self::Type) -> Result<Option<Self::Type>> {
+        Ok(Some(*arg))
+    }
+}
+
+pub struct RoundReal;
+
+impl Round for RoundReal {
+    type Type = Real;
+
+    #[inline]
+    fn round(arg: &Self::Type) -> Result<Option<Self::Type>> {
+        Ok(Some(Real::from(arg.round())))
+    }
+}
+
+pub struct RoundDec;
+
+impl Round for RoundDec {
+    type Type = Decimal;
+
+    #[inline]
+    fn round(arg: &Self::Type) -> Result<Option<Self::Type>> {
+        let result: codec::Result<Decimal> = arg
+            .to_owned()
+            .round(DEFAULT_FSP, RoundMode::HalfEven)
```

Comment:
Here we might need handle overflow with an `EvalContext`.

Reply:
Okay